.github/workflows/format.yml

```
@@ -0,0 +1,47 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - '**'
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  ci:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          persist-credentials: true
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+
+      - name: Apply rustfmt
+        run: cargo fmt --all
+
+      - name: Commit & push formatting changes
+        if: github.event_name == 'push'
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git add .
+          # if there’s anything staged, commit & push
+          git diff --cached --quiet || \
+            (git commit -m "chore: apply rustfmt formatting [skip ci]" && git push)
+
+      - name: Check formatting
+        run: cargo fmt --all -- --check
+
+      - name: Run Clippy
+        run: cargo clippy --workspace -- -D warnings
```

This will also lint the tests:

```suggestion
        run: cargo clippy --workspace --all-targets -- -D warnings
```